### ISSUE TITLE
Clarify how to interface npm with a local Git private registry

### ISSUE THREAD

**Issue by mwpowellhtx** [npm/npm#19791]
It is not clear from the [documentation](http://docs.npmjs.com/cli/install), and not much clearer from the [test](/npm/npm/blob/5e426a78ca02d0044f8dd26e0c5f881217081cbd/test/tap/add-remote-git-file.js#L20), how this exactly works. Where is my ``child.git`` found, for instance?

**Comment by mwpowellhtx** [comment]
For my part, I commit my *tarball* and *git tag* it. Rinse and repeat for as many versions as I want to track in package form. But from an *npm* consumer stand point, how does one reference the file based repository using ``git+file``?

**Comment by mwpowellhtx** [comment]
I have an error report as follows. It looks as though npm is trying to interpret my path as ``C:\``, incorrectly so; however, ``G:\`` is the root where the repository actually is.

In my test ``package.json``:

```JSON
"devDependencies": {
    "extensionsjs-array":  "git+file://G:/path/to/my/repo/@1.2.3" 
}
```

```
0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Dev\\js\\nodejs\\8.9.4\\node.exe',
1 verbose cli   'C:\\Dev\\js\\nodejs\\8.9.4\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'install' ]
2 info using npm@5.6.0
3 info using node@v8.9.4
4 verbose npm-session 8c5925462505139a
5 silly install runPreinstallTopLevelLifecycles
6 silly preinstall extensionsjs-tests@1.0.0
7 info lifecycle extensionsjs-tests@1.0.0~preinstall: extensionsjs-tests@1.0.0
8 silly install loadCurrentTree
9 silly install readLocalPackageData
10 silly install loadIdealTree
11 silly install cloneCurrentTreeToIdealTree
12 silly install loadShrinkwrap
13 silly install loadAllDepsIntoIdealTree
14 silly fetchPackageMetaData error for extensionsjs-array@git+file://G:/path/to/my/repo/@1.2.3 Error while executing:
14 silly fetchPackageMetaData C:\Dev\Git\cmd\git.EXE ls-remote -h -t file://g/path/to/my/repo/@1.2.3
14 silly fetchPackageMetaData
14 silly fetchPackageMetaData fatal: 'C:/Dev/Git/path/to/my/repo/@1.2.3' does not appear to be a git repository
14 silly fetchPackageMetaData fatal: Could not read from remote repository.
14 silly fetchPackageMetaData
14 silly fetchPackageMetaData Please make sure you have the correct access rights
14 silly fetchPackageMetaData and the repository exists.
14 silly fetchPackageMetaData
14 silly fetchPackageMetaData exited with error code: 128
15 silly saveTree extensionsjs-tests@1.0.0
16 verbose stack Error: exited with error code: 128
16 verbose stack     at ChildProcess.onexit (C:\Dev\js\nodejs\8.9.4\node_modules\npm\node_modules\mississippi\node_modules\end-of-stream\index.js:39:36)
16 verbose stack     at emitTwo (events.js:126:13)
16 verbose stack     at ChildProcess.emit (events.js:214:7)
16 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
17 verbose cwd G:\Source\Kingdom Software\JavaScript\ExtensionsJS\Working\src\ExtensionsJS.TestPackaging
18 verbose Windows_NT 6.1.7601
19 verbose argv "C:\\Dev\\js\\nodejs\\8.9.4\\node.exe" "C:\\Dev\\js\\nodejs\\8.9.4\\node_modules\\npm\\bin\\npm-cli.js" "install"
20 verbose node v8.9.4
21 verbose npm  v5.6.0
22 error Error while executing:
22 error C:\Dev\Git\cmd\git.EXE ls-remote -h -t file://g/path/to/my/repo/@1.2.3
22 error
22 error fatal: 'C:/Dev/Git/path/to/my/repo/@1.2.3' does not appear to be a git repository
22 error fatal: Could not read from remote repository.
22 error
22 error Please make sure you have the correct access rights
22 error and the repository exists.
22 error
22 error exited with error code: 128
23 verbose exit [ 1, true ]
```

**Comment by YasharF** [comment]
I am trying to understand the use case.  Why would an npm consumer, other than the dev want to access a repo on the local drive?

**Comment by mwpowellhtx** [comment]
@YasharF I think you understand the question well.

Yes, the dev, and/or his team, want to access a package from local (private) registry. It's for dev purposes only, draft purposes, that sort of thing.

Thanks!

**Comment by YasharF** [comment]
I could be wrong, but could the error that you are getting be because of forward slash and backslash issues since this seems to be a git on windows?

If I create a bare git repo in CMD (aka at c:\temp\gitrepo), add it as a remote in another project (`git remote add origin file://c:\temp\gitrepo`) and run npm init, by default I end up with the following, which has quite a few `/`s and `\`s mixed up in the url.  It doesn't seem to produce any errors when running `npm install` and `git push` though.
```
  "repository": {
    "type": "git",
    "url": "file://c:\\\\temp\\\\gitrepo"
  },
```

**Comment by mwpowellhtx** [comment]
@YasharF I've used forward/back-slash interchangeably in other parts of my ``Gulpfile.js`` without issue. In fact, seems to be more consistent, better, with forward-slash. That said, I do not know how ``npm`` wants the path, or how many levels of backslash escape characters are necessary.

For instance, doesn't ``c:\\\\temp\\\\gitrepo`` seem a bit excessive? Should ``c:\\temp\\gitrepo`` be adequate?

**Comment by mwpowellhtx** [comment]
@YasharF I tried again with:

```JavaScript
"devDependencies": {
    "my-package": "git+file://G:\\\\path\\\\to\\\\my-package#1.2.3"
}
```

Which yields the same result.

I'm not sure what is meant by "repository"? In what context? I wouldn't want to change the repository for the entire consuming ``package.json``, only the package dependency of interest.

```
0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Dev\\js\\nodejs\\8.9.4\\node.exe',
1 verbose cli   'C:\\Dev\\js\\nodejs\\8.9.4\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'install' ]
2 info using npm@5.6.0
3 info using node@v8.9.4
4 verbose npm-session 3173f66fa9f16a62
5 silly install runPreinstallTopLevelLifecycles
6 silly preinstall extensionsjs-tests@1.0.0
7 info lifecycle extensionsjs-tests@1.0.0~preinstall: extensionsjs-tests@1.0.0
8 silly install loadCurrentTree
9 silly install readLocalPackageData
10 silly install loadIdealTree
11 silly install cloneCurrentTreeToIdealTree
12 silly install loadShrinkwrap
13 silly install loadAllDepsIntoIdealTree
14 silly fetchPackageMetaData error for extensionsjs-array@git+file://G:\\path\\to\\my-package#1.2.3 Error while executing:
14 silly fetchPackageMetaData C:\Dev\Git\cmd\git.EXE ls-remote -h -t file://g//path//to//my-package
14 silly fetchPackageMetaData
14 silly fetchPackageMetaData fatal: '//path//to//my-package' does not appear to be a git repository
14 silly fetchPackageMetaData fatal: Could not read from remote repository.
14 silly fetchPackageMetaData
14 silly fetchPackageMetaData Please make sure you have the correct access rights
14 silly fetchPackageMetaData and the repository exists.
14 silly fetchPackageMetaData
14 silly fetchPackageMetaData exited with error code: 128
15 silly saveTree extensionsjs-tests@1.0.0
16 verbose stack Error: exited with error code: 128
16 verbose stack     at ChildProcess.onexit (C:\Dev\js\nodejs\8.9.4\node_modules\npm\node_modules\mississippi\node_modules\end-of-stream\index.js:39:36)
16 verbose stack     at emitTwo (events.js:126:13)
16 verbose stack     at ChildProcess.emit (events.js:214:7)
16 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
17 verbose cwd G:\path\to\my-test-fixture
18 verbose Windows_NT 6.1.7601
19 verbose argv "C:\\Dev\\js\\nodejs\\8.9.4\\node.exe" "C:\\Dev\\js\\nodejs\\8.9.4\\node_modules\\npm\\bin\\npm-cli.js" "install"
20 verbose node v8.9.4
21 verbose npm  v5.6.0
22 error Error while executing:
22 error C:\Dev\Git\cmd\git.EXE ls-remote -h -t file://g//path//to//my-package
22 error
22 error fatal: '//path//to//my-package' does not appear to be a git repository
22 error fatal: Could not read from remote repository.
22 error
22 error Please make sure you have the correct access rights
22 error and the repository exists.
22 error
22 error exited with error code: 128
23 verbose exit [ 1, true ]
```

**Comment by YasharF** [comment]
I'm not sure about the local path based repo with versioning, but ...
> the dev, and/or his team, want to access a package from local (private) registry. It's for dev purposes only, draft purposes, that sort of thing.

Going back to your use case, a couple of "workarounds" to unblock you in case if you are blocked right now:
One is to just use a clone of the package locally; and for different versions just have the dev checkout the different branches.  I know that works when a depency is just as plain as:
```
  "dependencies": {
    "mypackage": "file:c:/temp/mypackage"
  }
```
Another one is to setup a git and/or npm server onsite.  You might even be able to use something like https://www.npmjs.com/package/sinopia to just have a dev npm registry hosted locally on the dev desktop and be done.

**Comment by mwpowellhtx** [comment]
@YasharF This is what is in question. What else does ``git+file`` protocol mean, then, if not "local repository"? At any rate, I tried that as well,  and I got the same sort of errors.

Yes, I am also considering ``sinopia`` (or based on it, I think, if memory serves), as a viable path forward. I just wasn't clear what the ``git-xyz`` protocols would support, but it is becoming clear that a local repository isn't one of those paths.

Unless something else pops up, at the moment, I am unclear whether that's the case. Of course, if that ever changed, that would be great.
